plugins/StereoControl/StereoControl.cpp

```diff
@@ -0,0 +1,444 @@
+			{
+				delayResult[i] = linearInterpolate(m_delayBuf[i][floor(readLoc)], m_delayBuf[i][ceil(readLoc)], readLocFrac);
+			}
+			else// For when the interpolation wraps around to the beginning of the buffer
+			{
+				delayResult[i] = linearInterpolate(m_delayBuf[i][m_delayBufSize - 1], m_delayBuf[i][0], readLocFrac);
+			}
+
+			// Increment delay ring buffer index
+			++m_delayIndex[i];
+			if (m_delayIndex[i] >= m_delayBufSize)
+			{
+				m_delayIndex[i] -= m_delayBufSize;
+			}
+			// Send new value to delay line
+			m_delayBuf[i][m_delayIndex[i]] = s[i];
+		}
+
+		for (int i = 0; i < 2; ++i)
+		{
+			float allpassValue = (delayResult[0] + delayResult[1]) * 0.5f;
+
+			for (int j = 0; j < 2; ++j)
+			{
+				allpassValue = calcAllpassFilter(allpassValue, m_sampleRate, j, i, m_filtA, m_filtB);
+			}
+
+			allpassValue *= i ? -1.f : 1.f;
+
+			float lp;
+			float hp;
+			float hp2;
+			multimodeFilter(allpassValue, multimodeCoeff(stereoizerHP), lp, hp, m_stereoizerFilter[0][i]);
+			multimodeFilter(hp, multimodeCoeff(stereoizerLP), lp, hp2, m_stereoizerFilter[1][i]);
+
+			s[i] += lp * qMin(stereoizer, 1.f);
+		}
+		// ^^^ Stereoizer ^^^
+
+		const float sMid = (s[0] + s[1]) * 0.5f;
+		const float sSide = (s[0] - s[1]) * 0.5f;
+
+		s[0] = sMid + sSide * width;
+		s[1] = sMid - sSide * width;
+
+		if (mono)
+		{
+			s[0] = (s[0] + s[1]) * 0.5f;
+			s[1] = s[0];
+		}
+
+		if (monoBass)
+		{
+			float lp[2];
+			float hp[2];
+
+			for (int i = 0; i < 2; ++i)
+			{
+				// We apply two filters because a single filter doesn't have a strong enough slope
+				multimodeFilter(s[i], multimodeCoeff(monoBassFreq), lp[i], hp[i], m_monoBassFilter[i][0]);
+				multimodeFilter(s[i], multimodeCoeff(monoBassFreq), lp[i], hp[i], m_monoBassFilter[i][1]);
+			}
+
+			float lowmono = (lp[0] + lp[1]) * 0.5f;
+
+			if (audition)
+			{
+				s[0] = lowmono;
+				s[1] = lowmono;
+			}
+			else
+			{
+				s[0] = lowmono - hp[0];// Subtraction is needed due to filter phase shift
+				s[1] = lowmono - hp[1];
+			}
+		}
+
+		switch (panMode)
+		{
+			case 0:// Gain
+			{
+				const float lGain = pan > 0 ? 1.f - pan : 1.f;
+				const float rGain = pan < 0 ? 1.f + pan : 1.f;
+				s[0] *= lGain;
+				s[1] *= rGain;
+				break;
+			}
+			case 1:// Stereo
+			{
+				const float lGainL = panDualL > 0 ? 1.f - panDualL : 1.f;
+				const float rGainL = panDualL < 0 ? 1.f + panDualL : 1.f;
+				const float lGainR = panDualR > 0 ? 1.f - panDualR : 1.f;
+				const float rGainR = panDualR < 0 ? 1.f + panDualR : 1.f;
+				const float temp = s[0];
+				s[0] = s[0] * lGainL + s[1] * lGainR;
+				s[1] = temp * rGainL + s[1] * rGainR;
+				break;
+			}
+			case 2:// Haas
+			{
+				float haasResult = 0;
+				float haasDelayVal[2] = {0};
+
+				if (pan >= 0)
+				{
+					haasDelayVal[0] = linearInterpolate(0, 0.0008f * m_sampleRate * panDelay, abs(pan));
+				}
+				else
+				{
+					haasDelayVal[1] = linearInterpolate(0, 0.0008f * m_sampleRate * panDelay, abs(pan));
+				}
```

Comment:
- Why did you use linear interpolation instead of direct multiplication?
- I guess 0.8ms(`0.0008f` here) is approximately the distance between two ears divided by the speed of sound wave. Theoretically, the time difference should equal that value multiplied by `sin(angle)` where `angle` is the deviation of the source to left/right. This means `arcsin(panDelay * abs(pan))` will work like the deviation angle, right?

Reply:
Forgot to mention this: please consider extracting `0.0008f` to a named variable to make the code more readable.

Reply:
"Why did you use linear interpolation instead of direct multiplication?"
Because sometimes I program past midnight when I really should be sleeping.
"[Other stuff]"
Oh wow.  I'll trust you then.  The value I chose was mostly chosen semi-arbitrarily (which probably means I mercilessly stole it from another plugin, most likely Auburn's Psypan which this was heavily inspired by).